### PR TITLE
add middleware implementation

### LINKED ISSUE #72: Use middleware to replace OrchestrationRunner.loadAndRun in orchestrator functions
See https://github.com/Azure/azure-functions-java-worker/issues/595 for reference.

The idea is to simplify the orchestrator function programming model from this:

```java
@FunctionName("Chaining")
public String helloCitiesOrchestrator(
        @DurableOrchestrationTrigger(name = "runtimeState") String runtimeState) {
    return OrchestrationRunner.loadAndRun(runtimeState, ctx -> {
        String input = ctx.getInput(String.class);
        int x = ctx.callActivity("F1", input, int.class).await();
        int y = ctx.callActivity("F2", x, int.class).await();
        int z = ctx.callActivity("F3", y, int.class).await();
        return  ctx.callActivity("F4", z, double.class).await();
    });
}
```

...to something that looks like this...

```java
@FunctionName("Chaining")
public double helloCitiesOrchestrator(
        @DurableOrchestrationTrigger(name = "runtimeState") String input) {
    int x = ctx.callActivity("F1", input, int.class).await();
    int y = ctx.callActivity("F2", x, int.class).await();
    int z = ctx.callActivity("F3", y, int.class).await();
    return  ctx.callActivity("F4", z, double.class).await();
}
```

### PR DESCRIPTION
**Pull request by kaibocai**
### Issue describing the changes in this PR

resolves #72

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes are added to the `CHANGELOG.md`
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                if (cause instanceof OrchestratorBlockedEvent){
+                    throw (OrchestratorBlockedEvent) cause;
+                }
+                throw new RuntimeException("Unexpected failure in the task execution", e);
+            }
+        });
+        context.setMiddlewareOutput(orchestratorOutput);
+    }
+
+    private static Pair isOrchestrationTrigger(MiddlewareContext context) {
```
Comment (trask) [inline azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java:48]:
```suggestion
    private static Parameter isOrchestrationTrigger(MiddlewareContext context) {

```

File: azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package com.microsoft.durabletask.azurefunctions.internal.middleware;
```
Comment (cgillum) [inline azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java:1]: File needs copyright header.

File: azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ * at any time.
+ */
+public class OrchestrationMiddleware implements FunctionWorkerMiddleware {
+
+    private static final String ORCHESTRATION_TRIGGER = "DurableOrchestrationTrigger";
+
+    @Override
+    public void invoke(MiddlewareContext context, FunctionWorkerChain next) throws Exception {
+        Pair pair = isOrchestrationTrigger(context);
+        if (!pair.isOrchestrationTrigger){
```
Comment (cgillum) [inline azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java:27]: I think this logic needs to be refactored a bit to make the intent more clear. Perhaps it should look something like this:
```java
TriggerInfo info = getTriggerInfo(context)
if (!info.isOrchestrationTrigger) {
```

File: samples-azure-functions/src/main/java/com/functions/AzureFunctions.java [hunk 4]
```diff
@@ -35,17 +34,29 @@ public HttpResponseMessage startOrchestration(
      * This is the orchestrator function. The OrchestrationRunner.loadAndRun() static
      * method is used to take the function input and execute the orchestrator logic.
      */
+//    @FunctionName("Cities")
+//    public String citiesOrchestrator(
+//            @DurableOrchestrationTrigger(name = "orchestratorRequestProtoBytes") String orchestratorRequestProtoBytes) {
+//        return OrchestrationRunner.loadAndRun(orchestratorRequestProtoBytes, ctx -> {
```
Comment (cgillum) [inline samples-azure-functions/src/main/java/com/functions/AzureFunctions.java:40]: If we're going to keep this code here, can we add comments explaining that this is what an orchestrator function looks like during the preview, prior to GA?

Comment (kaibocai) [reply samples-azure-functions/src/main/java/com/functions/AzureFunctions.java:40]: Hi Chris, sorry for the confusion, I am still reviewing my change and will updates and clean up the code once the change is ready.

Comment (cgillum) [reply samples-azure-functions/src/main/java/com/functions/AzureFunctions.java:40]: Ah, okay - no problem. I'll wait until I see the request to review. 😊

**Review by cgillum — APPROVED** [review]
LGTM!

File: azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public class OrchestrationMiddleware implements Middleware {
+
+    private static final String ORCHESTRATION_TRIGGER = "DurableOrchestrationTrigger";
+
+    @Override
+    public void invoke(MiddlewareContext context, MiddlewareChain next) throws Exception {
```
Comment (trask) [inline azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java:26]:
```suggestion
    public void invoke(MiddlewareContext context, MiddlewareChain chain) throws Exception {

```

File: samples-azure-functions/src/main/java/com/functions/AzureFunctions.java [hunk 6]
```diff
@@ -37,15 +36,14 @@ public HttpResponseMessage startOrchestration(
...
-            @DurableOrchestrationTrigger(name = "orchestratorRequestProtoBytes") String orchestratorRequestProtoBytes) {
-        return OrchestrationRunner.loadAndRun(orchestratorRequestProtoBytes, ctx -> {
-            String result = "";
-            result += ctx.callActivity("Capitalize", "Tokyo", String.class).await() + ", ";
-            result += ctx.callActivity("Capitalize", "London", String.class).await() + ", ";
-            result += ctx.callActivity("Capitalize", "Seattle", String.class).await() + ", ";
-            result += ctx.callActivity("Capitalize", "Austin", String.class).await();
-            return result;
-        });
+            @DurableOrchestrationTrigger(name = "taskOrchestrationContext") TaskOrchestrationContext ctx)
```
Comment (trask) [inline samples-azure-functions/src/main/java/com/functions/AzureFunctions.java:39]: in future, consider either using "DurableOrchestration" everywhere or just "Orchestration" everywhere

File: azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        String parameterName = context.getParameterName(ORCHESTRATION_TRIGGER);
+        if (parameterName == null){
+            next.doNext(context);
+            return;
+        }
+        String orchestratorRequestProtoBytes = (String) context.getParameterValue(parameterName);
+        String orchestratorOutput  = OrchestrationRunner.loadAndRun(orchestratorRequestProtoBytes, ctx -> {
```
Comment (trask) [inline azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java:33]:
```suggestion
        String orchestratorOutputEncodedProtoBytes  = OrchestrationRunner.loadAndRun(orchestratorRequestProtoBytes, taskOrchestrationContext -> {

```

File: azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    private static final String ORCHESTRATION_TRIGGER = "DurableOrchestrationTrigger";
+
+    @Override
+    public void invoke(MiddlewareContext context, MiddlewareChain next) throws Exception {
+        String parameterName = context.getParameterName(ORCHESTRATION_TRIGGER);
+        if (parameterName == null){
+            next.doNext(context);
+            return;
+        }
+        String orchestratorRequestProtoBytes = (String) context.getParameterValue(parameterName);
```
Comment (trask) [inline azurefunctions/src/main/java/com/microsoft/durabletask/azurefunctions/internal/middleware/OrchestrationMiddleware.java:32]:
```suggestion
        String orchestratorRequestEncodedProtoBytes = (String) context.getParameterValue(parameterName);

```

**Review by trask — APPROVED** [review]

**Pull request merged**